pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+
+					if cmvt.validator.NumResolvedWithRows >= requestedResolved {
+						break
+					}
+				}
+			}
+
+			close(cmvt.verifierDone)
+			return nil
+		})
+	}
+}
+
+// validatorVerify checks if the validator has found any issues at the
+// time the function is called.
+func (cmvt *cdcMixedVersionTester) validatorVerify() versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		if failures := cmvt.validator.Failures(); len(failures) > 0 {
+			t.Fatalf("validator failures:\n%s", strings.Join(failures, "\n"))
+		}
+	}
+}
+
+// createChangeFeed issues a call to the given node to create a change
+// feed for the warehouse table.
+func (cmvt *cdcMixedVersionTester) createChangeFeed(node int) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("creating changefeed")
+		db := u.conn(ctx, t, node)
+		cdcClusterSettings(t, sqlutils.MakeSQLRunner(db))
+
+		opts := []string{"updated", "resolved"}
+		if _, err := db.Exec(
+			fmt.Sprintf("CREATE CHANGEFEED FOR %s INTO $1 WITH %s", target, strings.Join(opts, ", ")),
+			cmvt.kafka.sinkURL(ctx),
+		); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func runCDCMixedVersions(
+	ctx context.Context, t test.Test, c cluster.Cluster, buildVersion version.Version,
+) {
+	predecessorVersion, err := PredecessorVersion(buildVersion)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tester := newCDCMixedVersionTester(ctx, t, c)
+	tester.StartKafka(t, c)
+	defer tester.Cleanup()
+
+	// sleep is a step added at different points in the test when we
+	// want the workload to have some time to run against a specific
+	// version of the database
+	sleep := sleepStep(10 * time.Second)
+
+	// An empty string will lead to the cockroach binary specified by flag
+	// `cockroach` to be used.
+	const mainVersion = ""
+	newVersionUpgradeTest(c,
+		uploadAndStartFromCheckpointFixture(tester.crdbNodes, predecessorVersion),
+		tester.setupVerifier(1),
+		tester.installAndStartTPCCWorkload("15m"),
+		waitForUpgradeStep(tester.crdbNodes),
+
+		// NB: at this point, cluster and binary version equal predecessorVersion,
+		// and auto-upgrades are on.
+		preventAutoUpgradeStep(1),
+		tester.createChangeFeed(1),
+
+		// let the workload run in the old version for a while
+		sleep,
+		// Roll the nodes into the new version one by one in random order
+		binaryUpgradeStep(tester.crdbNodes, mainVersion),
+		// let the workload run in the new version for a while
+		sleep,
+
+		tester.validatorVerify(),
+
+		// Roll back again, which ought to be fine because the cluster upgrade was
+		// not finalized.
+		binaryUpgradeStep(tester.crdbNodes, predecessorVersion),
+		sleep,
+
+		tester.validatorVerify(),
+
+		// Roll nodes forward and finalize upgrade.
+		binaryUpgradeStep(tester.crdbNodes, mainVersion),
+
+		// allow cluster version to update
+		allowAutoUpgradeStep(1),
+		waitForUpgradeStep(tester.crdbNodes),
+
+		tester.waitForVerifier(),
```

Comment:
Would it make sense to fold `waitForVerifier` inside `validatorVerify`? I can't seem to come up with a use-case for waiting without validation.